design/core_collection_operations.md

```diff
@@ -0,0 +1,175 @@
+* There may be many versions of collections as they are updated and adjusted over time.
+  * Including adding new data products to collections.
+* Data product data structures / UIs / etc for collections will usually, but not always,
+  be applicable to other collections as well.
+
+## Collection operations / versions
+
+* Save a collection version.
+  * Administration only operation.
+  * Once saved, a collection version is immutable and cannot be overwritten.
+    * Trying to save a collection with an existing version will result in an error.
+  * A collection version *is not* immediately active.
+    * At some point we may wish to supply admin endpoints for accessing alternate collection
+      versions.
+  * The user specifies the version, which to the collections API is an opaque string. Any
+    semantics are supplied by the user.
+    * Note that this means semantic versions will not sort in order - the DB and API would need
+      specific schemas / code to do that.
+* Activate a collection version.
+  * Administration only operation.
+  * Can activate any version at any time.
+  * The active version is the only version accessible to users.
+  * Separating the saving and activation steps means we can save a collection document,
+    load data products for that collection, and then activate the collection when the data
+    products are ready.
+  * It also means we can trivially roll back changes.
+* Get a collection version.
+  * Public operation.
+  * Initially, this is only the active collection.
+    * At some point we may wish to supply admin endpoints for accessing alternate collection
+      versions.
+* List collections
+  * Public operation
+  * Only lists active collections.
+* Longer term:
+  * List all collections, even those without active versions.
+    * Administration only operation.
+    * For now, can be done relatively easily via the Aardvark Arango UI.
+  * List collection versions
+    * Administration only operation.
+    * This may be a long list. Returning only the list of version strings for one collection 
+      is probably reasonable up to ~100K versions, which seems much larger than we'll ever
+      encounter.
+
+## Collection contents
+
+For a minimal start, more will be added later (e.g. data comparison methods, contributors, etc.)
+
+* ID (lowercase alpha only string, no punctuation other than underscores, otherwise opaque)
+* Name (opaque string)
+* Version (opaque string)
+* Source version (opaque string)
+  * The source version may be the same from collection version to collection version if the
+    same source data is reloaded to, say, fix bugs or add features.
+  * This implies a particular collection version should not mix data from different source
+    versions.
+* Icon url
+  * Could probably use FastAPI's built in static server to serve these for now
+* Creation date
+* (If active) activation date
+* List of available data products with the active version of each data product (see example
```

Comment:
can we define "data product"? Is this the language we are going to use for the "widgets"/visualizations going forward

Reply:
i.e. is a "data product" the schema/api for each of those views, or is a "data product" the cached data sent/consumed by the api/ui?

Reply:
Data product is the vocabulary I've been using, but I'm not attached to it at all and if there's better terminology I'm all for it.

> i.e. is a "data product" the schema/api for each of those views, or is a "data product" the cached data sent/consumed by the api/ui?

Yes to both; I think of it as the data generated from the source data that supports the UI, the schema for that data in Arango, the API, and the UI

Reply:
I.e. the full stack for a particular widget